Dockerfile

```diff
@@ -0,0 +1,25 @@
+FROM node:6-stretch
+LABEL maintainer="will.price94@gmail.com"
+LABEL version="0.0.1"
+RUN apt-get update -qq && \
+    apt-get install -y python
+# Prevent npm from spamming
+RUN npm config set loglevel=warn && \
+    npm config set progress=false
+WORKDIR /app/
+COPY package.json package-lock.json /app/
+RUN npm install
+COPY . /app/
```

Comment:
probably a remnant of when you tried to make the build smaller, but is redundant for now

Reply:
No, this is critical, this is copying the source code *into* the container.

Reply:
sorry for being unclear.  I meant to say that `COPY . /app/` could replace that `COPY package.json package-lock.json /app/`

Reply:
This could be `COPY . .`

Reply:
mostly i just mean the two COPY lines can just be one

Reply:
Technically yes, practially no. I explicitly designed the Dockerfile this way to make use of the layer caching. Each line in the dockerfile creates a new image layer, if we copy everything in before running `npm install` then each time you make a source code change it will invalidate the layer that contains all the dependencies causing a subsequent rebuild to repull the dependencies. By only copying the package json files in and running `npm install` BEFORE adding the source code we only invalidate the layer when `package.json` changes.

Reply:
ah, I understand.  so every command corresponds to a layer, by default?  sounds good then

Reply:
"By default"? Well yes, but it's more than that, layers are a fundamental way of how docker works, it was designed that way so that things that don't change frequently are cached. If you look at other dockerfiles, all the good ones will copy dependency information in before copying source code in to take advantage of layers, this is good docker practice.